### PR TITLE
Nisp 1803 - Adding MQP logic in the backend

### PR DESCRIPTION
**Pull request by roundrobin99**
This is as far as I can take it. 

There is some debugging code in the console output to display the scenario been selected, so you can easily test the different scenarios.

### CODE REVIEW

**Comment by dspork** [comment]
This is not what I expected.
I expected a completely different field added to the ForecastModel in the SPSummaryModel (an Optional MQP Enum).
I would have expected then for you to write a another mqpScenario function in the forecasting service and unit test it and integrate it into the response.

Please remove any debugging code before submitting a PR, if there is unit tests it's not required.

**Comment by roundrobin99** [comment]
A scenario is a scenario (MQP or otherwise) so why do we need an Enum ? Isn't it better to use to the existing code base and extend it as I have done ? Anyway I submitted the PR for review mostly, as I know there will be changes so i left the debugging code in for the time being.

**Comment by dspork** [comment]
@roundrobin99 The MQP Scenario is independent of the forecasting Scenario
* A MQP user can be ContinueWorkingMax
* A MQP user can be ContinueWorkingNonMax
* A MQP user can be FillGaps

If you want to continue with the one enum we'll need to edit the Scenario Enum as follows:
  val Reached = Value
  val ContinueWorkingMax = Value
  val ContinueWorkingNonMax = Value
  val FillGaps = Value
  val ForecastOnly = Value
  val CantGetPension = Value
  val CanGetWithGaps = Value

We then add an MQP flag to backend response so we can check the scenario.

File: app/uk/gov/hmrc/nisp/services/ForecastingService.scala [hunk 1]
```diff
@@ -172,7 +174,15 @@ trait ForecastingService {
       case (false, false) => Scenario.ContinueWorkingNonMax
     }
   }
-
+  def MqpScenario(currentAmount: SPAmountModel, forecastAmount: SPAmountModel, personalMaximumAmount: SPAmountModel,
```
Comment (dspork) [inline app/uk/gov/hmrc/nisp/services/ForecastingService.scala]: The casing is wrong on the definition and this function needs Unit Tests.

File: test/uk/gov/hmrc/nisp/services/ForecastingServiceSpec.scala [hunk 2]
```diff
@@ -377,7 +377,7 @@ class ForecastingServiceSpec extends UnitSpec with OneAppPerSuite {
 
     val testNino = TestAccountBuilder.regularNino
     val hc = HeaderCarrier()
-
+/**
```
Comment (dspork) [inline test/uk/gov/hmrc/nisp/services/ForecastingServiceSpec.scala]: Why  is this test commented out now?

File: app/uk/gov/hmrc/nisp/services/ForecastingService.scala (hunk 1 above)
Comment (roundrobin99) [inline app/uk/gov/hmrc/nisp/services/ForecastingService.scala]: Do you mean the `case` conditions or the code casing ?

Comment (dspork) [inline app/uk/gov/hmrc/nisp/services/ForecastingService.scala]: I mean it should be `mqpScenario`

File: test/uk/gov/hmrc/nisp/services/ForecastingServiceSpec.scala (hunk 2 above)
Comment (roundrobin99) [inline test/uk/gov/hmrc/nisp/services/ForecastingServiceSpec.scala]: I forgot about it!

**Comment by roundrobin99** [comment]
I was thinking of passing the MQP scenario as you describe i.e MQPContinueWorkingMax or MQPContinueWorkingNonMax, doing as little processing in the frontend as possible. Logically it is part of the forecast scenario that way, isnt't it ?

**Comment by dspork** [comment]
I'm unsure about tightly coupling the Forecast Scenarios and MQP users, because it means duplicating scenarios and segregating the ContinueWorking users.

I guess we have 3 options.

1. A list of all scenarios including MQP ones
2. A list of the unique scenarios as above ^ with an MQP flag.
3. Leave the forecasting scenarios as they are and create the optional MQP Enum field.

Let's ask @prashantchoudhari

**Comment by roundrobin99** [comment]
I guess the better way would be to provide an MQP enum field. I will work on this now then , unless @prashantchoudhari has any objections.

**Comment by prashantchoudhari** [comment]
I think what @dspork is saying regarding (not tight coupling) MQP is right.

**Pull request closed without merging**